In [3]:
import cvxpy as cp
import numpy as np

In [7]:
products = ['obsidian', 'cells', 'thermalite', 'lava cake', 'magma ink', 'scoria paste', 'ashes', 'volcanic incense', 'sulfur reactor']
man_rets = [0.03, 0.30, 0.40, -0.50, 0.10, 0.0, -0.05, -0.30, 0]
rets = np.array(man_rets)
pi = cp.Variable(9)
objective = cp.Maximize((1_000_000 * rets.T @ pi) - (cp.sum_squares(pi) * 1_000_000))
constraints = [cp.norm(pi, 1) <= 100]
prob = cp.Problem(objective, constraints)

prob.solve()
print('Optimal allocation without integer constraints:')
sum = 0
fees = 0
profit = 0
for i in range(9):
    print("Position in ", products[i], ': ', f"{pi.value[i]:,.2f}", '%', ' | Given expected move: ', man_rets[i], sep='')
    sum += abs(pi.value[i])
    fees += (pi.value[i]**2)*1_000_000
    profit += (pi.value[i])*1_000_000*man_rets[i]

print(f"Total allocation: {sum}")
print(fees)
print(profit-fees)

Polishing not needed - no active set detected at optimal point
Optimal allocation without integer constraints:
Position in obsidian: 0.02% | Given expected move: 0.03
Position in cells: 0.15% | Given expected move: 0.3
Position in thermalite: 0.20% | Given expected move: 0.4
Position in lava cake: -0.25% | Given expected move: -0.5
Position in magma ink: 0.05% | Given expected move: 0.1
Position in scoria paste: 0.00% | Given expected move: 0.0
Position in ashes: -0.03% | Given expected move: -0.05
Position in volcanic incense: -0.15% | Given expected move: -0.3
Position in sulfur reactor: 0.00% | Given expected move: 0
Total allocation: 0.8400058743956379
150851.82541221983
150849.99999300626


| Product | Expected move | Position |
| -------- | ------- | ------- |
| Obsidian  | 0.03 | 0.02% |
| Pyro Cells | 0.3 | 0.15% |
| Thermalite | 0.4 | 0.20% |
| Lava Cake | -0.5 | 0.20% |
| Magma Ink | 0.1 | 0.20% |
| Scoria Pas. | 0.0 | 0.20% |
| Ashes | -0.05 | 0.20% |
| Volc. Incense | -0.3 | 0.20% |
| Sulfur Reactor | 0.0 | 0.20% |
| **Spent** |  | ~840k |
| **Fees** |  | ~150k |
| **Profit** |  | ~150k |
